SSR: Add support for media, sizes and heights attribute

sebastianbenz opened the pull request:
This will make it possible to remove the boilerplate even if the `media`, `sizes` or `heights` attributes are used. These attributes will be transformed into CSS media queries instead.

sebastianbenz commented:
@westonruter @schlessera PTAL - I haven't yet implemented the CSS size check.

westonruter commented on packages/optimizer/lib/transformers/ApplyCommonAttributes.js:
```diff
@@ -0,0 +1,254 @@
+
+const parseSizes = require('../parseSizes');
+const {appendChild, createElement, insertText, hasAttribute} = require('../NodeUtils');
+const ID_PREFIX = 'i-amp-';
```
The prefix `i-amphtml-` is disallowed, but `i-amp-` is OK. It seems somewhat strange why `i-amp-` wasn't chosen rather for the internal prefix since it has fewer bytes.

westonruter commented on packages/optimizer/lib/transformers/ApplyCommonAttributes.js:
```diff
@@ -0,0 +1,254 @@
+  toString() {
+    let result = '';
+    for (const {heights, defaultHeight, id} of this.heights) {
+      const selector = `#${id}:first-child`;
```
Shouldn't this be:

```suggestion
      const selector = `#${id}>:first-child`;
```

See https://github.com/ampproject/amp-wp/pull/4482#discussion_r421605106

westonruter commented on packages/optimizer/lib/transformers/ApplyCommonAttributes.js:
```diff
@@ -0,0 +1,254 @@
+      appendChild(head, customStyles);
+      insertText(customStyles, '');
+    }
+    customStyles.children[0].data += styles;
```
As noted in the description, we'll need to make sure this doesn't go above 75000 bytes.

westonruter commented on packages/optimizer/lib/transformers/ApplyCommonAttributes.js:
```diff
@@ -0,0 +1,254 @@
+      return node.attribs.id;
+    }
+    node.attribs = node.attribs || [];
+    node.attribs.id = ID_PREFIX + this.counter;
```
Just in case…

```suggestion
    node.attribs.id = ID_PREFIX + this.counter.toString();
```

schlessera commented on packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_heights_attribute_to_css/input.html:
```diff
@@ -2,32 +2,27 @@
     <script async src=https://cdn.ampproject.org/v0.js></script>
     <script async custom-element=amp-img src=https://cdn.ampproject.org/v0/amp-img-0.1.js></script>
+    <noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
     <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
 -moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes-amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes-amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes-amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyfra ...
-    <noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
```
Why the change in order here?

sebastianbenz replied:
reverted

schlessera commented on packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_media_attribute_to_css/expected_output.html:
```diff
@@ -0,0 +1,28 @@
+<head><style amp-runtime></style>
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+  <script async custom-element="amp-img" src="https://cdn.ampproject.org/v0/amp-img-0.1.js"></script>
+
+
```
Removal of the boilerplate shouldn't leave empty lines behind.

sebastianbenz replied:
that's going to be removed by a different transformer (not used in the test)

sebastianbenz replied:
This will be fixed by a different transformer

schlessera commented on packages/optimizer/lib/parseSizes.js:
```diff
@@ -0,0 +1,72 @@
+ * See https://html.spec.whatwg.org/multipage/images.html#parsing-a-sizes-attribute
+ *
+ * @param {String} string
+ * @returns {!Sizes}
```
Why the "non-nullable" modifier `!` here specifically? Wouldn't that be the default without the modifier anyways? It seems like that distinction isn't use anywhere else.

sebastianbenz replied:
It's a closure compiler flag used by the AMP runtime. Removed it as we don't need it.

sebastianbenz commented on packages/optimizer/lib/transformers/ApplyCommonAttributes.js:
```diff
@@ -0,0 +1,254 @@
+
+  transform(node, id) {
+    // normalize whitespace
+    let mediaString = node.attribs.media.replace(/\s+/g, ' ');
```
parseSizes is unrelated to the media attribute or am I missing something?

schlessera replied:
Oh, nvm, I was thinking of the media conditions that are part of the `sizes` attribute.

sebastianbenz commented on packages/optimizer/lib/transformers/ApplyCommonAttributes.js:
```diff
@@ -0,0 +1,254 @@
+   *
+   * @param {Node} node
+   */
+  getOrCreateId(node) {
```
True. This is not straightforward to implement in our case unfortunately.

sebastianbenz replied:
I'll ignore this case for now. Re-running the transformer would simply re-use the existing id. Pre-existing Ids should not be avoided by our prefix.

sebastianbenz replied:
Done

sebastianbenz commented:
I've addressed the comments. In regards to checking for the CSS limit. I'm currently lending towards not including this check:

* CSS optimization would need to be run twice (before, to be able to get the correct byte count and after, to optimize the generated CSS rules).
* I'd prefer to fail on not being able to remove the boilerplate, the perf implications of not being able to remove the boilerplate because of this are a lot harder to detect otherwise.

sebastianbenz commented:
@westonruter @schlessera PTAL

westonruter commented:
>  In regards to checking for the CSS limit. I'm currently lending towards not including this check

Doesn't this run the risk of the Optimizer generating invalid AMP from valid AMP input? If the CSS is close to the limit and then optimization causes it to go over, that seems not ideal?

sebastianbenz commented:
Merging as there's no further feedback that needs to be addressed.

schlessera commented:
@sebastianbenz There seems to be an issue with the CSS in here that was raised twice but always resolved without a comment, so I assume that GitHub resolved it because of an unrelated change: https://github.com/ampproject/amp-toolbox/pull/763#discussion_r425816316

Or am I missing something maybe?

sebastianbenz commented:
Hah! Not sure how this got lost. Fixed in #811

sebastianbenz commented:
Thanks for picking up on this!
